File: src/main/java/com/amazonaws/iot/fleetwise/vehiclesimulator/cert/ACMCertificateManager.java
package com.amazonaws.iot.fleetwise.vehiclesimulator.cert;

import com.amazonaws.iot.fleetwise.vehiclesimulator.DeviceCertificateConfig;
import com.amazonaws.iot.fleetwise.vehiclesimulator.SimulationMetaData;
import com.amazonaws.iot.fleetwise.vehiclesimulator.VehicleSetupStatus;
import com.amazonaws.iot.fleetwise.vehiclesimulator.storage.S3Storage;
import kotlin.text.Charsets;
import org.bouncycastle.asn1.x500.X500Name;
import org.bouncycastle.operator.ContentSigner;
import org.bouncycastle.operator.OperatorCreationException;
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
import org.bouncycastle.pkcs.PKCS10CertificationRequest;
import org.bouncycastle.pkcs.jcajce.JcaPKCS10CertificationRequestBuilder;
import org.bouncycastle.util.io.pem.PemObject;
import org.bouncycastle.util.io.pem.PemWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.acmpca.AcmPcaAsyncClient;
import software.amazon.awssdk.services.acmpca.model.*;
import software.amazon.awssdk.services.acmpca.waiters.AcmPcaAsyncWaiter;

import java.io.IOException;
import java.io.StringWriter;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class ACMCertificateManager {
    private final AcmPcaAsyncClient acmPcaClient;
    private final S3Storage s3Storage;

    public static final int CREATE_CERT_BATCH_SIZE = 10;

    private final Logger log = LoggerFactory.getLogger(ACMCertificateManager.class);

    public ACMCertificateManager(AcmPcaAsyncClient acmPcaClient, S3Storage s3Storage){
        this.acmPcaClient = acmPcaClient;
        this.s3Storage = s3Storage;
    }

    public VehicleSetupStatus setupVehiclesWithCertAndPrivateKey(List<SimulationMetaData> simConfigMap) {
        {
            Set<String> successVehicles = new HashSet<>();
            AtomicInteger counter = new AtomicInteger();
            Map<Integer, List<SimulationMetaData>> listOfChunks = simConfigMap.stream()
                    .collect(Collectors.groupingBy(it -> counter.getAndIncrement() / CREATE_CERT_BATCH_SIZE));
            listOfChunks.forEach((key, value) -> {
                List<CompletableFuture<String>> jobs = new ArrayList<>();
                value.forEach(
                        simInput -> {
                            CompletableFuture<String> cf =
                                    CompletableFuture.supplyAsync(() ->
                                    {
                                        DeviceCertificateConfig deviceCertConfig = simInput.getDeviceCertificateConfig();
                                        log.info("Creating private key and device certificate for vehicle: {}", simInput.getVehicleId());
                                        CertificateAndPrivateKey certAndPrivateKey = createDeviceCertWithCaCertAndPrivateKey(
                                                deviceCertConfig.getPcaArn(),
                                                deviceCertConfig.getCommonName(),
                                                deviceCertConfig.getOrganization(),
                                                deviceCertConfig.getCountryCode(),
                                                deviceCertConfig.getValidityDays()
                                        );
                                        s3Storage.put(simInput.getS3().getBucket(), simInput.getS3().getKey() + "/cert.crt", certAndPrivateKey.getCertificatePem().getBytes());
                                        s3Storage.put(simInput.getS3().getBucket(), simInput.getS3().getKey() + "/pri.key", certAndPrivateKey.getPrivateKey().getBytes());
                                        return simInput.getVehicleId();
                                    });
                            jobs.add(cf);
                        }
                );
                jobs.forEach(job -> {
                            try {
                                successVehicles.add(job.get());
                            } catch (InterruptedException | ExecutionException e) {
                                throw new RuntimeException(e);
                            }
                        }
                );
            });
            Set<String> failedVehicles = simConfigMap.stream().map(SimulationMetaData::getVehicleId).collect(Collectors.toSet());
            failedVehicles.removeAll(successVehicles);
            return new VehicleSetupStatus(successVehicles, failedVehicles);
        }
    }

    private CertificateAndPrivateKey createDeviceCertWithCaCertAndPrivateKey(String caArn,
            String commonName,
             String organization,
             String countryCode,
             Long expirationDurationDays) {
        if(expirationDurationDays == null){
            expirationDurationDays = 120L;
        }
        // create a private key pair
        KeyPairGenerator keyPairGenerator = null;
        try {
            keyPairGenerator = KeyPairGenerator.getInstance("RSA");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        keyPairGenerator.initialize(2048);
        KeyPair keyPair = keyPairGenerator.generateKeyPair();
        log.info("Created Private and Public Key Pair using RSA algorithm with 2048 bits.");

        // create a certificate signing request
        log.info("Creating CSR for device certificate");
        String csr = createCsr(keyPair, commonName, organization, countryCode);
        SdkBytes csrBytes = SdkBytes.fromByteArray(csr.getBytes(Charsets.UTF_8));

        log.info("CSR Generated. Issuing device certificate with CA ARN: {}", caArn);
        CompletableFuture<IssueCertificateResponse> issueCertificateResponseFuture = acmPcaClient.issueCertificate(
                IssueCertificateRequest.builder()
                        .certificateAuthorityArn(caArn)
                        .csr(csrBytes)
                        .signingAlgorithm(SigningAlgorithm.SHA256_WITHRSA)
                        .validity(
                                Validity.builder()
                                        .value(expirationDurationDays)
                                        .type(ValidityPeriodType.DAYS)
                                        .build()
                        )
                        .build()
        );
        IssueCertificateResponse issueCertificateResponse = null;
        try {
            issueCertificateResponse = issueCertificateResponseFuture.get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }

        String deviceCertArn = issueCertificateResponse.certificateArn();
        log.info("Waiting for device certificate to be issued, with ARN: {}", deviceCertArn);

        AcmPcaAsyncWaiter waiter = acmPcaClient.waiter();
        GetCertificateRequest getCertificateRequest = GetCertificateRequest.builder()
                .certificateAuthorityArn(caArn)
                .certificateArn(deviceCertArn)
                .build();

        try {
            waiter.waitUntilCertificateIssued(getCertificateRequest).get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
        log.info("Device Certificate has been issued, with ARN: {}", deviceCertArn);

        String deviceCert = null;
        try {
            deviceCert = acmPcaClient.getCertificate(getCertificateRequest).get().certificate();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
        log.info("Device Certificate has been retrieved, with ARN: {}", deviceCertArn);

        log.info("Retrieving CA Certificate with CA ARN: {}", caArn);
        String caCert = "";
        try {
            caCert = acmPcaClient.getCertificateAuthorityCertificate(
                    GetCertificateAuthorityCertificateRequest.builder()
                            .certificateAuthorityArn(caArn)
                            .build()
            ).get().certificate();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
        log.info("CA Certificate has been retrieved, with CA ARN: {}", caArn);

        // Concatenate device cert and ca cert in a single file
        String certificatePem = deviceCert +"\n" + caCert;
        log.info("Returning device certificate with CA Certificate. Number of bytes: {}", certificatePem.length());

        StringWriter stringWriter = new StringWriter();
        PemObject pemObject = new PemObject("RSA PRIVATE KEY", keyPair.getPrivate().getEncoded());
        PemWriter pemWriter = new PemWriter(stringWriter);
        try {
            pemWriter.writeObject(pemObject);
            pemWriter.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return new CertificateAndPrivateKey(stringWriter.toString(),certificatePem);
    }

    private String createCsr(
            KeyPair keyPair,
            String commonName,
            String organization,
            String countryCode)
    {
        String input = "CN=" + commonName+ ", O=" + organization + ", C=" +countryCode;
        JcaPKCS10CertificationRequestBuilder csrBuilder = new JcaPKCS10CertificationRequestBuilder(
                new X500Name(input),
                keyPair.getPublic()
        );

        // Sign the CSR
        JcaContentSignerBuilder csBuilder = new JcaContentSignerBuilder("SHA256withRSA");
        ContentSigner contentSigner = null;
        try {
            contentSigner = csBuilder.build(keyPair.getPrivate());
        } catch (OperatorCreationException e) {
            throw new RuntimeException(e);
        }
        PKCS10CertificationRequest csr = csrBuilder.build(contentSigner);

        // Convert CSR to PEM format
        StringWriter stringWriter = new StringWriter();
        PemObject pemObject = null;
        try {
            pemObject = new PemObject("CERTIFICATE REQUEST", csr.getEncoded());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        PemWriter pemWriter = new PemWriter(stringWriter);
        try {
            pemWriter.writeObject(pemObject);
            pemWriter.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return stringWriter.toString();
    }

}

File: src/main/kotlin/com/amazonaws/iot/vehiclesimulator/cert/ACMCertificateManager.kt
package com.amazonaws.iot.fleetwise.vehiclesimulator.cert

import com.amazonaws.iot.fleetwise.vehiclesimulator.SimulationMetaData
import com.amazonaws.iot.fleetwise.vehiclesimulator.VehicleSetupStatus
import com.amazonaws.iot.fleetwise.vehiclesimulator.storage.S3Storage
import kotlinx.coroutines.async
import kotlinx.coroutines.supervisorScope
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder
import org.bouncycastle.pkcs.jcajce.JcaPKCS10CertificationRequestBuilder
import org.bouncycastle.util.io.pem.PemObject
import org.bouncycastle.util.io.pem.PemWriter
import org.slf4j.LoggerFactory
import software.amazon.awssdk.core.SdkBytes
import software.amazon.awssdk.services.acmpca.AcmPcaAsyncClient
import software.amazon.awssdk.services.acmpca.model.GetCertificateAuthorityCertificateRequest
import software.amazon.awssdk.services.acmpca.model.GetCertificateRequest
import software.amazon.awssdk.services.acmpca.model.IssueCertificateRequest
import software.amazon.awssdk.services.acmpca.model.SigningAlgorithm
import software.amazon.awssdk.services.acmpca.model.Validity
import software.amazon.awssdk.services.acmpca.model.ValidityPeriodType
import java.io.StringWriter
import java.security.KeyPair
import java.security.KeyPairGenerator

/**
 * Certificate and private key in PEM format
 */
data class CertificateAndPrivateKey(
    val privateKey: String,
    val certificatePem: String
)

class ACMCertificateManager(
    private val acmPcaClient: AcmPcaAsyncClient,
    private val s3Storage: S3Storage
) {

    private val log = LoggerFactory.getLogger(ACMCertificateManager::class.java)

    suspend fun setupVehiclesWithCertAndPrivateKey(
        simConfigMap: List<SimulationMetaData>
    ): VehicleSetupStatus = supervisorScope {
        val successVehicles = mutableSetOf<String>()

        simConfigMap.chunked(CREATE_CERT_BATCH_SIZE).forEach { simInputList ->
            val jobs = simInputList.map { simInput ->
                async {
                    // certificate code goes here
                    val deviceCertConfig = simInput.deviceCertificateConfig!!

                    log.info("Creating private key and device certificate for vehicle: ${simInput.vehicleId}")
                    val certAndPrivateKey = createDeviceCertWithCaCertAndPrivateKey(
                        deviceCertConfig.pcaArn,
                        deviceCertConfig.commonName,
                        deviceCertConfig.organization,
                        deviceCertConfig.countryCode,
                        deviceCertConfig.validityDays
                    )
                    s3Storage.put(simInput.s3.bucket, "${simInput.s3.key}/cert.crt", certAndPrivateKey.certificatePem.toByteArray())
                    s3Storage.put(simInput.s3.bucket, "${simInput.s3.key}/pri.key", certAndPrivateKey.privateKey.toByteArray())
                    simInput.vehicleId
                }
            }

            jobs.forEach { job ->
                try {
                    successVehicles.add(job.await())
                } catch (ex: Exception) {
                    log.error("Failed to create certificate", ex)
                }
            }
        }
        return@supervisorScope VehicleSetupStatus(successVehicles, simConfigMap.map { it.vehicleId }.toSet() - successVehicles)
    }

    fun createDeviceCertWithCaCertAndPrivateKey(
        caArn: String,
        commonName: String,
        organization: String,
        countryCode: String,
        expirationDurationDays: Long = 120
    ): CertificateAndPrivateKey {

        // create a private key pair
        val keyPairGenerator = KeyPairGenerator.getInstance("RSA")
        keyPairGenerator.initialize(2048)
        val keyPair = keyPairGenerator.generateKeyPair()
        log.info("Created Private and Public Key Pair using RSA algorithm with 2048 bits.")

        // create a certificate signing request
        log.info("Creating CSR for device certificate")
        val csr = createCsr(keyPair, commonName, organization, countryCode)
        val csrBytes = SdkBytes.fromByteArray(csr.toByteArray(Charsets.UTF_8))

        log.info("CSR Generated. Issuing device certificate with CA ARN: $caArn")
        val issueCertificateResponse = acmPcaClient.issueCertificate(
            IssueCertificateRequest.builder()
                .certificateAuthorityArn(caArn)
                .csr(csrBytes)
                .signingAlgorithm(SigningAlgorithm.SHA256_WITHRSA)
                .validity(
                    Validity.builder()
                        .value(expirationDurationDays)
                        .type(ValidityPeriodType.DAYS)
                        .build()
                )
                .build()
        ).get()

        val deviceCertArn = issueCertificateResponse.certificateArn()
        log.info("Waiting for device certificate to be issued, with ARN: $deviceCertArn")

        val waiter = acmPcaClient.waiter()
        val getCertificateRequest: GetCertificateRequest? = GetCertificateRequest.builder()
            .certificateAuthorityArn(caArn)
            .certificateArn(deviceCertArn)
            .build()

        waiter.waitUntilCertificateIssued(getCertificateRequest).get()
        log.info("Device Certificate has been issued, with ARN: $deviceCertArn")

        val deviceCert = acmPcaClient.getCertificate(getCertificateRequest).get().certificate()
        log.info("Device Certificate has been retrieved, with ARN: $deviceCertArn")

        log.info("Retrieving CA Certificate with CA ARN: $caArn")
        val caCert = acmPcaClient.getCertificateAuthorityCertificate(
            GetCertificateAuthorityCertificateRequest.builder()
                .certificateAuthorityArn(caArn)
                .build()
        ).get().certificate()
        log.info("CA Certificate has been retrieved, with CA ARN: $caArn")

        // Concatenate device cert and ca cert in a single file
        val certificatePem = "$deviceCert\n$caCert"
        log.info("Returning device certificate with CA Certificate. Number of bytes: ${certificatePem.length}")

        val stringWriter = StringWriter()
        val pemObject = PemObject("RSA PRIVATE KEY", keyPair.private.encoded)
        val pemWriter = PemWriter(stringWriter)
        pemWriter.writeObject(pemObject)
        pemWriter.close()

        return CertificateAndPrivateKey(
            privateKey = stringWriter.toString(),
            certificatePem = certificatePem
        )
    }

    fun createCsr(
        keyPair: KeyPair,
        commonName: String,
        organization: String,
        countryCode: String,
    ): String {
        val csrBuilder = JcaPKCS10CertificationRequestBuilder(
            X500Name("CN=$commonName, O=$organization, C=$countryCode"),
            keyPair.public
        )

        // Sign the CSR
        val csBuilder = JcaContentSignerBuilder("SHA256withRSA")
        val contentSigner = csBuilder.build(keyPair.private)
        val csr = csrBuilder.build(contentSigner)

        // Convert CSR to PEM format
        val stringWriter = StringWriter()
        val pemObject = PemObject("CERTIFICATE REQUEST", csr.encoded)
        val pemWriter = PemWriter(stringWriter)
        pemWriter.writeObject(pemObject)
        pemWriter.close()
        return stringWriter.toString()
    }

    companion object {
        const val CREATE_CERT_BATCH_SIZE = 10
    }
}
